File: python/src/days/day7.py
import io
from pathlib import Path
from days.day import Day

class Day7(Day):
    
    def process(self, input_filename: str, is_part2: bool = False) -> str:
        try:
            file_path = Path("resources") / input_filename
            with open(file_path, "r", encoding="utf-8") as reader:
                return parse_input(reader, is_part2)

        except FileNotFoundError as e:
            print(f"Error: Input file '{input_filename}' not found.")
            return "Error"
        except Exception as e:
            print(f"An error occurred: {e}")
            return "Error"


def parse_input(reader, is_part2):
    lines = [line.strip().split() for line in reader]
    # lines = []
    # for line in reader:
    #     lines.append(line.strip().split())

    result = calibration_helper(lines, is_part2)
    return str(result)


def calibration_helper(lines, is_part2):
    result = 0
    for line in lines:
        value = int(line[0][:-1])
        if is_valid_test(line, 1, 0, value, is_part2):
            result += value
    return result


def is_valid_test(line, pos, exp_value, value, is_part2):
    if pos == len(line):
        return exp_value == value

    curr_value = int(line[pos])

    if is_valid_test(line, pos + 1, exp_value + curr_value, value, is_part2):
        return True

    if is_valid_test(line, pos + 1, curr_value if exp_value == 0 else exp_value * curr_value, value, is_part2):
        return True

    if is_part2 and pos <= len(line) - 1:
        concatenated_value = int(str(exp_value) + str(curr_value))
        if is_valid_test(line, pos + 1, concatenated_value, value, is_part2):
            return True

    return False

File: java/src/days/Day7.java
package src.days;

import java.io.BufferedReader;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;

public class Day7 extends Day {
    @Override
    public String process(String inputFileName, boolean isPart2) {        
        Path filePath = Path.of("resources", inputFileName);
         try (BufferedReader reader = Files.newBufferedReader(filePath, StandardCharsets.UTF_8)) {        
            return parseInput(reader, isPart2);
        } catch (IOException e) {
            e.printStackTrace();
        }
        return "Error";
    }

    private static String parseInput(BufferedReader reader, boolean isPart2) throws IOException {
        String line;
        long ret = 0;
        ArrayList<String[]> lines = new ArrayList<>();

        while ((line = reader.readLine()) != null) {
            String[] input = line.split("\\s+");
            lines.add(input);            
        }

        ret = calibrationHelper(lines, isPart2);                    

        return String.valueOf(ret);
    }

    private static long calibrationHelper(ArrayList<String[]> lines, boolean isPart2) {
        long ret = 0;
        for(String[] line : lines) {
            long value = Long.parseLong(line[0].replace(":", ""));
            if(isValidTest(line,1,0, value, isPart2)) ret += value;
        }
        return ret;
    }


    private static boolean isValidTest(String[] line, int pos, long exp_value, long value, boolean isPart2) {
        if(pos == line.length && value == exp_value){
            return true;
        }

        if(pos == line.length) return false;
        long curr_value = Long.parseLong(line[pos]);
        
        if(isValidTest(line, pos + 1, exp_value + curr_value, value, isPart2)) return true;
        if(isValidTest(line, pos + 1, (exp_value == 0) ? curr_value : exp_value * curr_value, value, isPart2)) return true;

        if(isPart2 && pos <= line.length - 1)
            if(isValidTest(line, pos + 1, Long.parseLong(exp_value + "" + curr_value), value, isPart2)) return true; 
        
        
        return false;        

    }
}
